### PR TITLE
function toStartOfSecond(DateTime64) -> DateTime64

### PR DESCRIPTION
**Pull request by Enmk**
This function nullifies the sub-second part of DateTime64 value.

I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):
- New Feature

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
New function function toStartOfSecond(DateTime64) -> DateTime64 that nullifies sub-second part of DateTime64 value.

Detailed description / Documentation draft:
## toStartOfSecond(`value`, `timezone`)

Returns new DateTime64 with nullified sub-second part and (optionally) different timezone. 

* `value` - `DateTime64` 
* `timezone` - `String`, optional timezone for the return value.
* `return` - a `DateTime64` of the same precision (and same timezone unless specified explicitly) as `value`, but with sub-second part nullified.

Example:
``` sql
WITH toDateTime64('2019-09-16 19:20:11.123', 3) AS dt64 SELECT toStartOfSecond(dt64);
```

``` text
┌───toStartOfSecond(dt64)─┐
│ 2019-09-16 19:20:11.000 │
└─────────────────────────┘
```
...

### CODE REVIEW

File: src/Core/DecimalFunctions.h [hunk 1]
```diff
@@ -121,32 +126,41 @@ DecimalComponents<typename DecimalType::NativeType> split(const DecimalType & de
...
+template <typename DecimalType>
+inline typename DecimalType::NativeType getFractionalPartWithScaleMultiplier(
+        const DecimalType & decimal,
+        typename DecimalType::NativeType scale_multiplier)
+{
+    using T = typename DecimalType::NativeType;
+
+    T result = decimal.value;
+    if (result < T(0))
+        result *= T(-1);
```
Comment (alexey-milovidov) [inline src/Core/DecimalFunctions.h:147]: Maybe result = -result will look better?
(no difference in the behaviour indeed)

File: src/Functions/DateTimeTransforms.h [hunk 2]
```diff
@@ -175,6 +175,28 @@ struct ToStartOfMinuteImpl
...
 };
 
+struct ToStartOfSecondImpl
+{
+    static constexpr auto name = "toStartOfSecond";
+
+    static inline DateTime64 execute(const DateTime64 & datetime64, Int64 scale_multiplier, const DateLUTImpl &)
+    {
+        const auto fractional = DecimalUtils::getFractionalPartWithScaleMultiplier(datetime64, scale_multiplier);
+        return datetime64 - fractional;
```
Comment (alexey-milovidov) [inline src/Functions/DateTimeTransforms.h:185]: It does not look correct for negative values (that will be introduced in another PR).

Example:

value is -0.3
fractional part that you calculate is 0.3
value - fractional = -0.6

But if I understand correctly, you want to round down to -1.0 (towards minus infinity).

File: src/DataTypes/DataTypeDateTime64.h [hunk 3]
```diff
@@ -45,40 +45,75 @@ class DataTypeDateTime64 final : public DataTypeDecimalBase<DateTime64>, public
...
-class DateTime64BasicTransformWrapper : public Transform
+class TransformDateTime64 : public Transform
 {
+private:
+    // Name<X>::name is X::name or default.
+    template<typename X>
+    struct Name
+    {
+        template<typename U, typename CHECK> struct SFINAE {};
+        template<typename U> static constexpr auto getName(SFINAE<U, decltype(U::name)>*) { return U::name; }
```
Comment (alexey-milovidov) [inline src/DataTypes/DataTypeDateTime64.h:72]: It looks old-fashioned, maybe it's possible to use std::enable_if_t?

**Review by alexey-milovidov — CHANGES_REQUESTED** [review]
Small questions remain.

File: src/DataTypes/DataTypeDateTime64.h (hunk 3 above)
Comment (Enmk) [reply src/DataTypes/DataTypeDateTime64.h]: fixed

File: src/Functions/DateTimeTransforms.h (hunk 2 above)
Comment (Enmk) [reply src/Functions/DateTimeTransforms.h]: nice catch, fixed that

**Review by alexey-milovidov — APPROVED** [review]
